### PR TITLE
feat: Implement concurrent stat

### PR DESCRIPTION
**Pull request by Xuanwo**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Implement concurrent stat to speedup the list files.

fix https://github.com/datafuselabs/databend/issues/15082

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

### CODE REVIEW

File: src/common/storage/src/stage.rs [hunk 1]
```diff
@@ -247,6 +239,41 @@ impl StageFilesInfo {
...
+    pub async fn stat_concurrent(
+        &self,
+        operator: &Operator,
+        first_only: bool,
+        max_files: usize,
+        files: &Vec<String>,
+    ) -> Vec<Result<(String, Metadata)>> {
+        let mut futures = Vec::with_capacity(files.len());
+        let mut limit = 0;
+        for file in files {
```
Comment (BohuTANG) [inline src/common/storage/src/stage.rs:253]: It's better to use a batch; we parallelize the batch. I'm wondering if there are too many files, will there be too many futures?
See: https://github.com/datafuselabs/databend/blob/3a50f2c918b527ed140c4f83e36a0aac01d5b31c/src/query/storages/fuse/src/io/files.rs#L63-L77

Comment (Xuanwo) [reply src/common/storage/src/stage.rs:253]: Makes sense, let me change them into an iterator.

File: src/common/storage/src/stage.rs [hunk 2]
```diff
@@ -247,6 +240,46 @@ impl StageFilesInfo {
...
+                .to_string();
+            let operator = operator.clone();
+            async move {
+                let meta = operator.stat(&full_path).await?;
+                Ok((full_path, meta))
+            }
+        });
+
+        // We don't have access to ctx in current context, use 100 as a safe default value.
+        execute_futures_in_parallel(tasks, 100, 100 * 2, "batch-stat-file-worker".to_owned()).await
```
Comment (BohuTANG) [inline src/common/storage/src/stage.rs:281]: 100 is too large, It's better to set it to thread_num, please reference:
```
threads_nums, 
     threads_nums * 2, 
```

**Review by youngsofun — APPROVED** [review]

**Comment by Xuanwo** [comment]
Done, cc @BohuTANG PTAL

**Pull request merged**